### PR TITLE
feature: add redis resource access for PAM

### PR DESCRIPTION
**Pull request by fangpenlin**
## Context

ref: https://linear.app/infisical/issue/PAM-42/add-redis-account-type-for-pam

## Steps to verify the change

To test this feature, one needs to run a Redis server locally. First, let's see the example redis config file:

```
# Redis configuration file with multiple users

# Default user (admin) - requires password
requirepass admin123

# ACL Users configuration
# Format: user <username> on <password> <permissions> <keys>
# 
# Example users:
# 1. admin - full access
user admin on +@all ~* &* >admin123

# 2. appuser - read and write access
user appuser on +@read +@write ~* >apppass456

# 3. readonly - read-only access
user readonly on +@read ~* >readonly789

# 4. testuser - full access for testing
user testuser on +@all ~* >testpass321

# 5. developer - read, write, and some admin commands
user developer on +@read +@write +@keyspace +@string ~* >devpass999

# Bind to all interfaces
bind 0.0.0.0

# Enable protected mode (optional, but recommended)
protected-mode yes

```

With that, we can first have a redis server running without TLS:

```yaml
version: '3.8'

services:
  redis:
    image: redis:7-alpine
    container_name: redis
    restart: always
    ports:
      - "6378:6379"
    command: redis-server /usr/local/etc/redis/redis.conf
    volumes:
      - redis-non-tls-data:/data
      - ./redis.conf:/usr/local/etc/redis/redis.conf:ro
    healthcheck:
      test: ["CMD", "redis-cli", "-a", "admin123", "ping"]
      interval: 10s
      timeout: 3s
      retries: 5

volumes:
  redis-non-tls-data:

```

For the TLS version, you can first generate the certs with:

```bash
#!/bin/bash

mkdir -p certs && \
openssl genrsa -out certs/ca.key 4096 && \
openssl req -x509 -new -nodes -sha256 -key certs/ca.key -days 3650 -subj "/O=Redis Test/CN=Redis Test CA" -out certs/ca.crt && \
openssl genrsa -out certs/redis.key 2048 && \
openssl req -new -sha256 -key certs/redis.key -subj "/O=Redis Test/CN=Redis Server" -out certs/redis.csr && \
openssl x509 -req -sha256 -in certs/redis.csr -CA certs/ca.crt -CAkey certs/ca.key -CAcreateserial -days 365 \
  -extfile <(printf "[san]\nsubjectAltName=IP:192.168.50.215") -extensions san \
  -out certs/redis.crt && \
rm -f certs/redis.csr certs/ca.srl  # Clean up temporary files (use -f to avoid error if files don't exist)
```

then with the compose:

```yaml
version: '3.8'

services:
  redis-tls:
    image: redis:7-alpine
    ports:
      - "6377:6379"  # Host port 6378 maps to container TLS port 6379
    command:
      - redis-server
      - /usr/local/etc/redis/redis.conf
      - --port 0                  # Disable non-TLS port
      - --tls-port 6379           # Enable TLS on the standard Redis port inside container
      - --tls-cert-file /certs/redis.crt
      - --tls-key-file /certs/redis.key
      - --tls-ca-cert-file /certs/ca.crt
      - --tls-auth-clients no
    volumes:
      - redis-data:/data
      - ./redis.conf:/usr/local/etc/redis/redis.conf:ro
      - ./certs:/certs:ro         # Mount your certs directory (read-only)
    healthcheck:
      test: ["CMD", "redis-cli", "-a", "admin123", "--tls", "ping"]  # Update healthcheck for TLS (adjust certs if needed for internal check)
      interval: 10s
      timeout: 3s
      retries: 5

volumes:
  redis-data:
```

the testing process is pretty much largely the same as other PAM resources.

1. Create the Redis resource - go to PAM resource page and create a Redis server resource pointing to the redis server you want to test
2. Create a Redis account with credentials defined in the config file for the redis resource you want to test
3. Run CLI in branch https://github.com/Infisical/cli/pull/95 with the cmd generated for accessing the account
4. Run `redis -u <infisical cli proxy_url>` to test it. Some corner cases, such as pub/sub cmds, monitors are a bit special in terms of how our proxy handles them.
5. Kill the access cli process, check the session logs and see if they are uploaded

## Type

- [ ] Fix
- [x] Feature
- [ ] Improvement
- [ ] Breaking
- [ ] Docs
- [ ] Chore

### CODE REVIEW

**Comment by maidul98** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scanner |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/maidul98/pr-checks/192556e6-97ab-4f8c-a6ee-c22539380dce?refs=3a28a47486d77422a872301dd7a494bc05732b78&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
+import { FormProvider, useForm } from "react-hook-form";
+import { zodResolver } from "@hookform/resolvers/zod";
+import { z } from "zod";
+
+import { Button, ModalClose } from "@app/components/v2";
+import { PamResourceType, TRedisAccount } from "@app/hooks/api/pam";
+import { UNCHANGED_PASSWORD_SENTINEL } from "@app/hooks/api/pam/constants";
+
+import { SqlAccountFields } from "./shared/SqlAccountFields";
```
[the remark this answers is not in the document]
Comment (lb-vn) [reply frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx]: Agreed, even though they're both username + password combos, its best to separate them in case we expand SQL fields in the future. It'd break redis forms.

Also make the fields not be in /shared folder, since redis won't have any "branch resources" like SQL

File: backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+
+export interface RedisResourceConnection {
+  /**
+   * Check and see if the connection is good or not.
+   *
+   * @param connectOnly when true, if we only want to know that making the connection is possible or not,
+   *                    we don't care about authentication failures
+   * @returns Promise to be resolved when the connection is good, otherwise an error will be errbacked
+   */
+  validate: (connectOnly: boolean) => Promise<void>;
```
Comment (lb-vn) [inline backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts]: Do we even need `connectOnly` for `validate`? It does not support using auth for the Redis factory due to the new `authenticate` method.

File: backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+    } catch (error) {
+      if (error instanceof BadRequestError) {
+        if (
+          error.message.includes("NOAUTH") ||
+          error.message.includes("WRONGPASS") ||
+          error.message.includes("invalid password")
+        ) {
+          throw new BadRequestError({
+            message: "Account credentials invalid: Username or password incorrect"
+          });
```
Comment (lb-vn) [inline backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts:258]: This logic for detecting a wrong password is duplicated from the `validate` function above. Maybe we can turn this into some reusable method or get rid of the `authenticate` function and use `validate` with credentials similar to how SQL factory works?

File: backend/src/ee/services/pam-resource/redis/redis-resource-schemas.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    .string()
+    .trim()
+    .transform((value) => value || undefined)
+    .optional()
+});
+
+export const RedisAccountCredentialsSchema = z.object({
+  username: z.string().trim().min(1).max(63),
+  password: z.string().trim().min(1).max(256)
+});
```
Comment (lb-vn) [inline backend/src/ee/services/pam-resource/redis/redis-resource-schemas.ts:39]: Let's update username to be 256, since 63 is a Postgres limit.
Also don't `.trim()` Redis passwords since leading and trailing whitespace is valid.

File: frontend/src/hooks/api/pam/types/redis-resource.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { PamResourceType } from "../enums";
+import { TBasePamAccount } from "./base-account";
+import { TBasePamResource } from "./base-resource";
+
+export type TRedisConnectionDetails = {
+  host: string;
+  port: number;
+  username?: string;
+  password?: string;
```
Comment (lb-vn) [inline frontend/src/hooks/api/pam/types/redis-resource.ts]: Are these two really required? I don't think they're used

File: frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    username: z
+      .string()
+      .trim()
+      .min(1, "Username required")
+      .max(63, "Username must be 63 characters or less"),
+    password: z
+      .string()
+      .trim()
+      .min(1, "Password required")
+      .max(256, "Password must be 256 characters or less")
```
Comment (lb-vn) [inline frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx]: Username max must be 256, since 63 is a relic from postgres.
Also get rid of .trim() for password if possible since redis allows leading/trailing whitespace.

File: frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (lb-vn) [reply frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx]: Put the fields directly into this file, I don't see a need for a component

File: frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+  return (
+    <FormProvider {...form}>
+      <form
+        onSubmit={(e) => {
+          handleSubmit(onSubmit)(e);
+        }}
+      >
+        <GenericAccountFields />
+        <SqlAccountFields isUpdate={isUpdate} />
```
Comment (lb-vn) [inline frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx]: Like Greptile pointed out above, remove this SQL fields component and just create a username+password field here without a component

File: frontend/src/pages/pam/PamResourcesPage/components/PamResourceForm/shared/RedisResourceFields.tsx
Comment (lb-vn) [inline frontend/src/pages/pam/PamResourcesPage/components/PamResourceForm/shared/RedisResourceFields.tsx:1]: This file shouldn't be in the /shared folder since it's not a shared component. We can just get rid of this file and integrate this stuff directly into the `frontend/src/pages/pam/PamResourcesPage/components/PamResourceForm/RedisResourceForm.tsx` file

File: frontend/src/pages/pam/PamResourcesPage/components/PamResourceForm/RedisResourceForm.tsx [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        onSubmit={(e) => {
+          setSelectedTabIndex(0);
+          handleSubmit(onSubmit)(e);
+        }}
+      >
+        <GenericResourceFields />
+        <RedisResourceFields
+          selectedTabIndex={selectedTabIndex}
+          setSelectedTabIndex={setSelectedTabIndex}
+        />
```
Comment (lb-vn) [inline frontend/src/pages/pam/PamResourcesPage/components/PamResourceForm/RedisResourceForm.tsx]: Like I stated in a comment on the resource fields file, we can get rid of it from the /shared folder and just directly add the fields here since they're not reusable

File: frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (fangpenlin) [reply frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx]: fixed

File: backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts (hunk 2 above)
Comment (fangpenlin) [reply backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts]: good point. Now I've combined the two method into just one.to send `AUTH` or not will be determined whether if credentials are provided

File: backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts (hunk 3 above)
Comment (fangpenlin) [reply backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts:258]: fixed

File: backend/src/ee/services/pam-resource/redis/redis-resource-schemas.ts (hunk 4 above)
Comment (fangpenlin) [reply backend/src/ee/services/pam-resource/redis/redis-resource-schemas.ts:39]: fixed

File: frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx (hunk 7 above)
Comment (fangpenlin) [reply frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx]: fixed

File: frontend/src/hooks/api/pam/types/redis-resource.ts (hunk 5 above)
Comment (fangpenlin) [reply frontend/src/hooks/api/pam/types/redis-resource.ts]: they are not needed. removed

File: frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx (hunk 6 above)
Comment (fangpenlin) [reply frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx]: fixed

File: backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+          const result = await client.auth(credentials.username, credentials.password);
+          if (result !== "OK") {
+            throw new BadRequestError({
+              message: `Authentication failed: Redis returned ${result as string} status`
+            });
+          }
+        } else {
+          // Just validate connection with ping
+          await client.ping();
+        }
```
Comment (lb-vn) [inline backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts:104]: We can make this block cleaner by reusing the `isConnectOnly` variable. For example:
```
if isConnectOnly:
  await client.ping()
else:
  // ...
```

File: backend/src/ee/services/pam-resource/redis/redis-resource-schemas.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    .trim()
+    .transform((value) => value || undefined)
+    .optional()
+});
+
+export const RedisAccountCredentialsSchema = z.object({
+  username: z
+    .string()
+    .trim()
+    .max(256, "Username must be 255 characters or less")
```
Comment (lb-vn) [inline backend/src/ee/services/pam-resource/redis/redis-resource-schemas.ts]: nit: error says 255 but max is 256

File: frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+  resourceType?: PamResourceType;
+  onSubmit: (formData: FormData) => Promise<void>;
+};
+
+const formSchema = genericAccountFieldsSchema.extend({
+  credentials: z.object({
+    username: z
+      .string()
+      .trim()
+      .max(256, "Username must be 255 characters or less")
```
Comment (lb-vn) [inline frontend/src/pages/pam/PamAccountsPage/components/PamAccountForm/RedisAccountForm.tsx]: nit: error says 255 but max is 256

File: backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts (hunk 9 above)
Comment (fangpenlin) [reply backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts:104]: image
nope, unfortunately doesn't work. it seems like linter is not good at boolean algebra

Comment (lb-vn) [reply backend/src/ee/services/pam-resource/redis/redis-resource-factory.ts:104]: No problem, we'll skip this
